invenio_rdm_records/services/services.py

```diff
@@ -342,19 +342,86 @@ def delete_secret_link(
 
         return True
 
-    def get_pid(
-        self, id_, identity,
-    ):
-        """Reserve doi."""
-        #  provider =
-        # find_the_provider_by_the_param_value
-        # (<provider param value>)
-        #  return provider.reserve(...)
+    def get_provider(self, scheme, client=None):
+        """Process a given PID with its provider."""
+        provider_class = self.config.pids_providers[scheme]
+        if client:
+            provider = provider_class(
+                client=self.config.providers_clients[client]())
+        else:
+            provider = provider_class()
+
+        return provider
+
+    def reserve_pid(self, id_, identity, pid_type):
+        """Reserve PID for a given record."""
+        draft = self.draft_cls.pid.resolve(id_, registered_only=False)
 
-        return "1234"
+        # Permissions
+        # PIDS-FIXME: manage is the correct perm type?
+        self.require_permission(identity, "manage", record=draft)
 
-    def delete_pid(
-        self, id_, identity, doi,
-    ):
-        """Unreserve doi."""
-        return ""
+        # PIDS-FIXME: Where should client come from? URL?
+        provider = self.get_provider(pid_type, client="datacite")
+        if not provider:
+            # PIDS-FIXME: raise
+            return
+
+        pid = provider.create(id_)
+
+        # PIDS-FIXME: delegate to a minter?/component
+        draft.pids[pid_type] = {
+            "identifier": pid.pid_value,
+            "provider": provider.name
+        }
+        if provider.client:
+            draft.pids[pid_type]["client"] = provider._client_credentials.name
+
+        draft.commit()
```

Comment:
I am not sure I would this here. It might be that the user request a DOI, but then changes his/her mind and copy/paste another one, external. Wondering if you should simply reserve it in PIDStore, but not touch the draft.